pravega/pravega
# Issue 7104: Delayed future cancelation

**Change log description**  
Makes sure that delayed futures, async loops, and exponential backoff get canceled if the theread pool they are running on is shutdown, and a cancelation exception is propagated upwards.

**Purpose of the change**  
Fixes #7104
Also fixes #7327
Also fixes #7375 

**What the code does**  
Modifies the thread pool to supply a CompletableFuture to the caller of delay task so that this can be awaited rather than chaining another future off thread pool task. Similarly for Futures.loop and backoff the cancelation of the future and failure to submit another task to the pool are propagated to the result future.

Additionally: The style formatter had a pathological case when dealing with indentation of nested runnables as arguments as often occurs in async code. There is a new capability to deal with this, as such the format style is updated. (Due to the version update this is very noisy, but there is only one real change)

**How to verify it**  
New unit tests are added. However it should be run through system and other tests, as a sanity check because additional exceptions are being introduced which would have been silent before.
